DCOS-13159: Bug fix: Placement constraints

mlunoe opened the pull request:
Placement constraints was not working for pods and had a few issues.

This PR makes the placement constraints work for pods again with a separate reducer for this (there is a unique items issue that @wavesoft is working on separately). It also introduces form validation for placement constraints and displays `Not Applicable` in the ServiceConfigDisplay for `value`s that are empty.

NB: I encountered an error while working on this that was introduced here: https://github.com/dcos/dcos-ui/commit/31630d08344f2a2d00f2534a9333e084fc459a76

Before:
![](https://cl.ly/270S3P0Q1k2m/Image%202017-01-31%20at%2016.03.45.png)

After:
![](https://cl.ly/0v1R001j3O1b/Image%202017-01-31%20at%2016.32.54.png)

mlunoe commented:
Run integration tests

mlunoe commented:
This PR is missing some tests on the new validator that was introduced. It might need a similar validator for pods. Possibly add integration tests to check that pods can actually be created and edited using our new form (probably separate PR).

mlunoe commented:
Added some tests for the reducers. Still missing for the new validation util.

One integration is failing:
```
Service Actions
    Edit Action

      ��� opens the correct service edit modal

      1) closes modal on successful API request
```

Will try to take a look.

MatApple commented:
@rcorral @orlandohohmeier perhaps we could distribute ownership of @mlunoe's PRs so they get on the 1.9 train.

orlandohohmeier commented:
@nLight could you please take care?

nLight commented:
@MatApple @jfurrow rebased, added one more validator, added tests

nLight commented:
retest this please

orlandohohmeier commented:
![](https://cloud.githubusercontent.com/assets/647035/23170995/841a9ce2-f851-11e6-83bf-db2581fbcf7c.png)
I get the following error after creating pod with placement constraints. The declined offer util seems to have an issue dealing with the correct pod placement constraints format.

nLight commented:
@orlandohohmeier good catch! Fixed

orlandohohmeier commented on plugins/services/src/js/constants/OperatorTypes.js:
```diff
@@ -1,10 +1,28 @@
-  LIKE: 'LIKE',
-  UNLIKE: 'UNLIKE',
-  MAX_PER: 'MAX_PER'
+  types: {
```
I'd reorganize the config by operator to improve maintainability. 

```js
[
  {
    label: 'OPERATOR',
    type: 'OPERATOR',
    requiresVlaue: true
  }
]
```

nLight replied:
Done, but different ;)

orlandohohmeier commented on plugins/services/src/js/components/forms/GeneralServiceFormSection.js:
```diff
@@ -240,23 +240,36 @@ class GeneralServiceFormSection extends Component {
+      this.props.errors,
+      'constraints'
+    );
+    const showValueLabel = data.some((constraint) => {
```
Do we actually need `showValueLabel` and `showRequiredLabel` looks they always have the same values.

nLight replied:
removed

orlandohohmeier commented on plugins/services/src/js/components/forms/GeneralServiceFormSection.js:
```diff
@@ -240,23 +240,36 @@ class GeneralServiceFormSection extends Component {
+      );
 
       if (index === 0) {
         fieldLabel = this.getConstraintField(
```
Let's please use internal functional components instead methods like `getConstraintField` similar to the pod review screen.

nLight replied:
done

nLight commented:
As discussed offline I `hold merge` it to unify reducers because duplicating 120 LOC because of a few lines is too much.

nLight commented on plugins/services/src/js/constants/OperatorTypes.js:
```diff
@@ -1,10 +1,28 @@
 const OperatorTypes = {
```
It provides types and info whether they require value or not. I think the file name is ok.

nLight commented on plugins/services/src/js/components/forms/GeneralServiceFormSection.js:
```diff
@@ -190,24 +190,37 @@ class GeneralServiceFormSection extends Component {
+      'constraints'
+    );
+    const showValueLabel = constraints.some((constraint) => {
+      return hasThirdField[constraint.operator];
```
done

mlunoe commented:
Alright. Fixed a small issue I found and added a test for it.
This thing looks good! Let's get it merged! 🚢

mlunoe commented:
@jfurrow could you take a look?

jfurrow commented:
There seems to be an error with placement constraints that's not showing up until I get to the review screen. In the following screenshot, I corrected all errors in the form, then went to the review screen, then went back to the form:
![](https://cl.ly/1T422z2G291T/Screen%20Shot%202017-02-24%20at%209.43.35%20AM.png)

mlunoe commented:
Thank you @jfurrow for looking through this.

I updated the PR and submitted a few issues for Marathon: https://github.com/mesosphere/marathon/issues/5077#issuecomment-282435950 and https://github.com/mesosphere/marathon/issues/5253

Also, errors weren't really working for placement constraints. I changed this so even errors from Marathon will show up. Now when you try to submit, get an error from Marathon, it is displayed when you come back to the form.

I used these two app configs as training data:

Happy path:
```js
{
  "id": "/mlunoe/testing",
  "instances": 1,
  "container": {
    "type": "MESOS"
  },
  "cpus": 0.1,
  "mem": 128,
  "cmd": "sleep 100",
  "constraints": [
    [
      "foo",
      "UNIQUE"
    ],
    [
      "bar",
      "CLUSTER"
    ],
    [
      "baz",
      "GROUP_BY",
      "1"
    ],
    [
      "qux",
      "LIKE",
      "qux"
    ],
    [
      "quux",
      "UNLIKE",
      "quux"
    ],
    [
      "corgly",
      "MAX_PER",
      "2"
    ]
  ]
}
```

Error path:
```js
{
  "id": "/mlunoe/testing",
  "instances": 1,
  "container": {
    "type": "MESOS"
  },
  "cpus": 0.1,
  "mem": 128,
  "cmd": "sleep 100",
  "constraints": [
    [
      "foo",
      "UNIQUE"
    ],
    [
      "bar",
      "CLUSTER"
      "bar"
    ],
    [
      "baz",
      "GROUP_BY",
      "asf"
    ],
    [
      "qux",
      "LIKE",
      "qux"
    ],
    [
      "quux",
      "UNLIKE"
    ],
    [
      "corgly",
      "MAX_PER",
      "asf"
    ]
  ]
}
```

mlunoe commented:
(failing because of this error unrelated to this PR. Fix is here: #1953)

mlunoe commented:
Rebased off of latest master, which contains fix

jfurrow commented on plugins/services/src/js/components/forms/GeneralServiceFormSection.js:
```diff
@@ -227,6 +233,15 @@ class GeneralServiceFormSection extends Component {
+              {constraintsErrors}
+            </FieldError>
+          </FormGroup>
+        </FormRow>
```
Sorry, I know we did this together but I just realized that if there is no error, then this empty row causes some additional margin to stack up. Could we conditionally render this entire row, dependent on the existence of the error?

mlunoe replied:
I removed the extra row. This what it looks like with an error message, which is exactly how I imagined it to be:
![](https://cl.ly/1z28203F0O0Q/Image%202017-02-24%20at%2019.08.49.png)

It also looks right without an error message showing.

jfurrow commented:
Hey @mlunoe thanks for addressing my concerns! I can verify #1 and #2 are fixed but #3 is still happening, where you can't actually remove placement constraints.

mlunoe commented:
You are all good. Apparently, the issue was introduced here: https://github.com/dcos/dcos-ui/pull/1933/files#diff-56de52bdaf825c1afa13152b35bd4f7cR121. That PR was not tested thoroughly enough, I guess. We need to be careful of these things slipping through the cracks last minute @dcos/frontend.
I re-added the appConfig check to the implementation. @wavesoft please check that it is behaving as it needs to. There aren't any tests for this currently. I'll let you take care of that.

I also removed the extra row. This what it looks like with an error message, which is exactly how I imagined it to be:
![](https://cl.ly/1z28203F0O0Q/Image%202017-02-24%20at%2019.08.49.png)

mlunoe commented:
Run integration tests

jfurrow commented:
Nice work @mlunoe, it works well! 🚢

nLight commented:
Do we want to validate extra params?
```
[
      "hostname",
      "UNIQUE",
      "123"
]
```

nLight commented:
We might want to go even further to validate this, maybe as a separate PR tho
```
  "constraints": [
    [
      [
        "bla",
        "LIKE",
        "blip"
      ]
    ]
],
```

mlunoe commented:
Added validation (with tests) for the extra params that are required to be empty

mlunoe commented:
Yes, separate PR for the latter scenario

nLight commented:
As we've addressed all the Orlando's comments in one way or another I'm merging it.
